Packet++/header/SomeIpLayer.h

```diff
@@ -0,0 +1,435 @@
+		RESPONSE = 0x80,
+		/** The Acknowledgment for RESPONSE(informational) */
+		RESPONSE_ACK = 0xC0,
+		/** The response containing an error */
+		ERRORS = 0x81,
+		/** Acknowledgment for ERROR(informational) */
+		ERROR_ACK = 0xC1,
+		/** A TP request expecting a response (even void) */
+		TP_REQUEST = 0x20,
+		/** A TP fire&forget request */
+		TP_REQUEST_NO_RETURN = 0x21,
+		/** A TP request of a notification/event callback expecting no response */
+		TP_NOTIFICATION = 0x22,
+		/** The TP response message */
+		TP_RESPONSE = 0xa0,
+		/** The TP response containing an error */
+		TP_ERROR = 0xa1,
+	};
+
+	/**
+	 * @brief Represents an SOMEIP protocol header
+	 *
+	 */
+#pragma pack(push, 1)
+	struct someiphdr
+	{
+		/** Service ID */
+		uint16_t serviceID;
+		/** Method ID. Most significant bit 0 when E2E communication. 1 when SOMEIP event */
+		uint16_t methodID;
+		/** Length. Also covers payload. Excludes serviceID, methodID and length field itself */
+		uint32_t length;
+		/** Client ID */
+		uint16_t clientID;
+		/** Session ID */
+		uint16_t sessionID;
+		/** Protocol Version */
+		uint8_t protocolVersion;
+		/** Interface Version */
+		uint8_t interfaceVersion;
+		/** Message Type */
+		uint8_t msgType;
+		/** Return Code */
+		uint8_t returnCode;
+
+		someiphdr()
+			: serviceID(0), methodID(0), length(0), clientID(0), sessionID(0), protocolVersion(0), interfaceVersion(0),
+			  msgType(0), returnCode(0)
+		{
+		}
+	};
+
+	static_assert(sizeof(someiphdr) == 16, "Wrong Size");
+#pragma pack(pop)
+
+	/**
+	 * A constructor that creates the layer from an existing packet raw data
+	 * @param[in] data A pointer to the raw data (will be casted to @ref someiphdr)
+	 * @param[in] dataLen Size of the data in bytes
+	 * @param[in] prevLayer A pointer to the previous layer
+	 * @param[in] packet A pointer to the Packet instance where layer will be stored in
+	 */
+	SomeIpLayer(uint8_t *data, size_t dataLen, Layer *prevLayer, Packet *packet)
+		: Layer(data, dataLen, prevLayer, packet)
+	{
+		m_Protocol = SomeIp;
+
+		m_PayloadPtr = nullptr;
+	}
+
+	/**
+	 * @brief Construct a new SomeIp Layer object
+	 *
+	 * @param[in] serviceID Service ID
+	 * @param[in] methodID Method ID
+	 * @param[in] clientID Client ID
+	 * @param[in] sessionID Session ID
+	 * @param[in] interfaceVersion Interface Version
+	 * @param[in] type Type of the message
+	 * @param[in] returnCode Return Code
+	 * @param[in] data A pointer to the raw data
+	 * @param[in] dataLen Size of the data in bytes
+	 * @param[in] dontCopy If set to true, the payload is not copied to the layer object. In this case the m_PayloadPtr
+	 * holds the reference to a data buffer. This option can be used to reduce the number of copies to generate packets.
+	 */
+	SomeIpLayer(uint16_t serviceID, uint16_t methodID, uint16_t clientID, uint16_t sessionID, uint8_t interfaceVersion,
+				MsgType type, uint8_t returnCode, const uint8_t *const data = nullptr, size_t dataLen = 0,
+				bool dontCopy = false);

```

Comment:
I think `data`, `dataLen` and `dontCopy` are redundant. Assuming we consider the payload as a separate layer, a layer shouldn't create additional layers. The user should create the `SomeIpLayer` and the payload layer separately

Reply:
i removed the dontCopy flag

Reply:
Since SOME/IP-SD and SOME/IP-TP will be separate layers in the next PRs, I think we can also remove `data` and `dataLen`. If users want to create multiple layers they can create a `SomeIpLayer` and on top of that they can create other `SomeIpLayer`/`SomeIpSdLayer`/`SomeIpTpLayer`

Reply:
i think the data and dataLen cannot be reduced because the SOME/IP Layer also contains a payload, see also https://github.com/seladb/PcapPlusPlus/pull/895#discussion_r968200981

Reply:
Yes, that is correct!